libqtile/core/manager.py

```diff
@@ -747,7 +747,10 @@ def unmanage(self, wid: int) -> None:
             if isinstance(c, base.Window):
                 # Put the group back on the window so hooked functions can access it.
                 c.group = group
-            hook.fire("client_killed", c)
+
+            # Do not trigger for internal windows
+            if not isinstance(c, base.Internal):
+                hook.fire("client_killed", c)
```

Comment:
I wonder if, as you point out in your commit message, we should have a hook.fire_client_hook, which checks these properties so we can enforce consistency across all the client_* hooks? That way if we do decide to change it for all of them, it's just one place.

That said, this series looks good in general, and I'm happy with whatever outcome you decide. Thanks for sending it.

Reply:
Sounds like the cleaner solution, how would this `fire_client_hook` be implemented? `fire_client_hook(self, event, client, *args, **kwargs)` would not work because not all hooks involving `Window` objects have the client as the first argument.

Would I just keep the signature from `fire(self, event, *args, **kwargs)` and look for an instance of `Internal` in the `args`?

I wonder if a "smart" approach, foregoing an additional function and instead inspecting either the event name or args, would introduce too much obscure complexity?